pkg/reader/ccip.go

```diff
@@ -439,29 +443,101 @@ func (r *ccipChainReader) Nonces(
+	feeComponents := make(map[cciptypes.ChainSelector]types.ChainFeeComponents, len(r.contractWriters))
+	for chain, chainWriter := range r.contractWriters {
+		feeComponent, err := chainWriter.GetFeeComponents(ctx)
+		if err != nil {
+			r.lggr.Errorw("failed to get chain fee components for chain %d: %w", chain, err)
+			continue
+		}
+		feeComponents[chain] = *feeComponent
 	}
+	return feeComponents
+}
 
-	eg := new(errgroup.Group)
-	gasPrices := make([]cciptypes.BigInt, len(chains))
-	for i, chain := range chains {
-		i, chain := i, chain
-		eg.Go(func() error {
-			gasPrice, err := r.contractWriters[chain].GetFeeComponents(ctx)
-			if err != nil {
-				return fmt.Errorf("failed to get gas price: %w", err)
-			}
-			gasPrices[i] = cciptypes.NewBigInt(gasPrice.ExecutionFee)
-			return nil
-		})
+func (r *ccipChainReader) GetWrappedNativeTokenPriceUSD(
+	ctx context.Context,
+	selectors []cciptypes.ChainSelector,
+) map[cciptypes.ChainSelector]cciptypes.BigInt {
+	// 1. Call chain's router to get native token address https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/Router.sol#L189:L191
+	// nolint:lll
+	// 2. Call FeeQuoter to get native tokens price  https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/FeeQuoter.sol#L229-L229
+	prices := make(map[cciptypes.ChainSelector]cciptypes.BigInt, len(selectors))
+	for _, chain := range selectors {
+		reader, ok := r.contractReaders[chain]
+		if !ok {
+			r.lggr.Warnw("contract reader not found", "chain", chain)
+			continue
+		}
+
+		var nativeTokenAddress types2.Account
+		err := reader.ExtendedGetLatestValue(
+			ctx,
+			consts.ContractNameRouter,
+			consts.MethodNameRouterGetWrappedNative,
+			primitives.Unconfirmed,
+			nil,
+			&nativeTokenAddress)
+
+		if err != nil {
+			r.lggr.Errorw("failed to get native token address", "chain", chain, "err", err)
+			continue
+		}
+
+		var price *big.Int
+		err = reader.ExtendedGetLatestValue(
+			ctx,
+			consts.ContractNameFeeQuoter,
+			consts.MethodNameFeeQuoterGetTokenPrices,
+			primitives.Unconfirmed,
+			map[string]any{
+				"token": nativeTokenAddress,
+			},
+			&price,
+		)
+		if err != nil {
+			r.lggr.Errorw("failed to get native token price", "chain", chain, "err", err)
+			continue
+		}
+		prices[chain] = cciptypes.NewBigInt(price)
 	}
+	return prices
+}
 
-	if err := eg.Wait(); err != nil {
-		return nil, err
+// GetChainFeePriceUpdate Read from Destination chain FeeQuoter latest fee updates for the provided chains.
+// It unpacks the packed fee into the ChainFeeUSDPrices struct.
+// nolint:lll
+// https://github.com/smartcontractkit/chainlink/blob/60e8b1181dd74b66903cf5b9a8427557b85357ec/contracts/src/v0.8/ccip/FeeQuoter.sol#L263-L263
+func (r *ccipChainReader) GetChainFeePriceUpdate(ctx context.Context, selectors []cciptypes.ChainSelector) map[cciptypes.ChainSelector]plugincommon.ChainFeeUpdate {
```

Comment:
Should we have an explicit param for the dest chain here? Or are we still not sure we want to follow that pattern? cc @winder

Reply:
Isn't that what selectors is?

Reply:
I believe `selectors` here are sources. Would be nice to be more explicit

Reply:
selectors are sources (which are dest chain selectors on the dest chain). You want to make the API clearer by providing the destChain @makramkd ?